crypto/evp_extra/p_hmac_asn1.c

```diff
@@ -66,9 +67,52 @@ static int hmac_size(OPENSSL_UNUSED const EVP_PKEY *pkey) {
 
 static void hmac_key_free(EVP_PKEY *pkey) {
   HMAC_KEY *key = pkey->pkey.ptr;
+  if (key != NULL) {
+    OPENSSL_free(key->key);
+  }
   OPENSSL_free(key);
 }
 
+static int hmac_set_key(EVP_PKEY *pkey, const uint8_t *priv, size_t len,
+                        OPENSSL_UNUSED const uint8_t *pubkey,
+                        OPENSSL_UNUSED size_t pubkey_len) {
+  if (pkey->pkey.ptr != NULL) {
+    return 0;
+  }
+
+  HMAC_KEY *key = HMAC_KEY_new();
+  if (key == NULL) {
+    return 0;
+  }
+
+  key->key = OPENSSL_memdup(priv, len);
+  key->key_len = len;
+  pkey->pkey.ptr = key;
+  return 1;
+}
+
+static int hmac_get_key(const EVP_PKEY *pkey, uint8_t *priv, size_t *len) {
+  HMAC_KEY *key = pkey->pkey.ptr;
+  if (key == NULL) {
+    return 0;
+  }
+
+  // The semantics of the EVP APIs are to return the length, if |priv| is NULL.
+  if (priv == NULL) {
+    *len = key->key_len;
```

Comment:
should we do a null check on `len` here, or can we always assume that to be a valid pointer?